pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1009 @@
+			// TrafficControl.
+			if err = c.uninstallTrafficControl(tcName, tcState); err != nil {
+				return err
+			}
+			// Delete the state of the deleted TrafficControl.
+			c.deleteTrafficControlState(tcName)
+			return nil
+		}
+		return err
+	}
+
+	// Get the TrafficControl state.
+	tcState, exists := c.getTrafficControlState(tcName)
+	// If the TrafficControl exists and corresponding state doesn't exist, create state for the TrafficControl.
+	if !exists {
+		tcState = c.newTrafficControlState(tcName, tc.Spec.Action, tc.Spec.Direction)
+	}
+
+	if tc.Spec.ReturnPort != nil {
+		// Get name of the return port.
+		returnPortName = c.getPortName(tc.Spec.ReturnPort)
+		// If the name is different from the cached name in the TrafficControl state, it could be caused by the return
+		// port update of the TrafficControl or the creation of the TrafficControl.
+		if returnPortName != tcState.returnPortName {
+			if tcState.returnPortName != "" {
+				// If the stale return port name cached in TrafficControl state is not empty, release the stale return port
+				// from the TrafficControl.
+				returnOFPort, portDeleted, err := c.releaseTrafficControlPort(returnPortName, tcName)
+				if err != nil {
+					return err
+				}
+				// If the stale return port is deleted, uninstall the return flow for the return port.
+				if portDeleted {
+					if err = c.ofClient.UninstallTrafficControlReturnPortFlow(returnOFPort); err != nil {
+						return err
+					}
+				}
+				tcUpdated = true
+			}
+			// Get or create the return port.
+			var returnOFPort uint32
+			if returnOFPort, exists, err = c.getOrCreateTrafficControlPort(tc.Spec.ReturnPort, returnPortName, tcName); err != nil {
+				return err
+			}
+			// If the return port is newly created, install a return flow for the return port.
+			if !exists {
+				if err = c.ofClient.InstallTrafficControlReturnPortFlow(returnOFPort); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	// Get name of the return port.
+	targetPortName = c.getPortName(&tc.Spec.TargetPort)
+	// If the name is different from the cached name in the TrafficControl state, it could be caused by the target port
+	// update of the TrafficControl or the creation of the TrafficControl.
+	if targetPortName != tcState.targetPortName {
+		if tcState.targetPortName != "" {
+			// If the stale return port name cached in TrafficControl state is not empty, release the stale target port
+			// from the TrafficControl.
+			_, _, err = c.releaseTrafficControlPort(tcState.targetPortName, tcName)
+			if err != nil {
+				return err
+			}
+			tcUpdated = true
+		}
+	}
+	// Get or create the target port.
+	targetOFPort, _, err := c.getOrCreateTrafficControlPort(&tc.Spec.TargetPort, targetPortName, tcName)
+	if err != nil {
+		return err
+	}
+
+	// Update the TrafficControl state.
+	tcState.targetPortName = targetPortName
+	tcState.returnPortName = returnPortName
```

Comment:
I think it should assign returnPortName to tcState once `getOrCreateTrafficControlPort` succeeds, otherwise cleanup will base on stale return port if getting targetPort fails?

Reply:
Done.